### PR TITLE
Move V1 CloudEvent implementation into own namespace

### PR DESCRIPTION
**Pull request by rubenrangel**
In order to decouple implementation details in systems, I created the `CloudEventInterface` as a starting point, and moved the V1 implementation to its own namespace with a V1 interface. Currently, the interface is producing a read-only view of an event. The only function in the base interface returns the spec version for the event, but more functions could be moved into it. I didn't see anything in the spec that said the other attributes would be present across all major versions of the spec, but let me know if I missed that.

There might be a generic way of retrieving attributes by name and we can add that functionality to the base interface at a later point.

### CODE REVIEW

**Comment by rubenrangel** [comment]
Converted to a draft while I edit the title and message. Pressed the return key accidentally :/

File: composer.json [hunk 1]
```diff
@@ -23,14 +20,17 @@
             "CloudEvents\\Tests\\": "tests/"
         }
     },
+    "require": {
```
Comment (rubenrangel) [inline composer.json:23]: Just moved to be closer to the other `require` section.

File: src/CloudEvents/CloudEvent.php [hunk 2]
```diff
@@ -1,82 +0,0 @@
-<?php
```
Comment (rubenrangel) [inline src/CloudEvents/CloudEvent.php:1]: Moved to https://github.com/cloudevents/sdk-php/pull/4/files#diff-7489e1f1e0cc7562d90dd81ea209e40fd2517ad0140c331ed345b97c8d753bd1. I kept the function signatures the same.

File: src/CloudEvents/V1/CloudEvent.php [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+    private string $type;
+    private ?string $dataContentType;
+    private ?string $dataSchema;
+    private ?string $subject;
+    private ?DateTimeInterface $time;
+
+    /** @var mixed|null */
+    private $data;
+
+    public function __construct(
```
Comment (rubenrangel) [inline src/CloudEvents/V1/CloudEvent.php:24]: I thought about having just the required properties be in the constructor, but thinking forward to PHP 8, I feel this constructor signature will play nicely with named parameters and property promotion.

File: composer.json [hunk 4]
```diff
@@ -23,14 +20,17 @@
...
     "require-dev": {
         "nunomaduro/collision": "^5.2",
         "phpunit/phpunit": "^9.5",
         "squizlabs/php_codesniffer": "^3.5"
     },
     "scripts": {
-        "lint": "./vendor/bin/phpcs --standard=PSR12 ./src",
-        "lint-fix": "./vendor/bin/phpcbf --standard=PSR12 ./src",
+        "lint": "./vendor/bin/phpcs --standard=PSR12 ./src ./tests",
+        "lint-fix": "./vendor/bin/phpcbf --standard=PSR12 ./src ./tests",
```
Comment (jlaswell) [inline composer.json:33]: Great catch to enforce this here.

**Review by jlaswell — APPROVED** [review]
The PR is looking really good! The only concern I have is the `CloudEvents\CloudEventsInterface` with a single method. The only benefit I see with a consumer is to have functionality based on the spec version of the provided Cloud event, which will always be V1 as of now. I think the V1 interface makes a lot of sense for something like a Laravel adapter to adapt "Laravel" events into CloudEvents because it has some encapsulation of behavior.

I can also see the idea of a consumer switching behavior based on the spec version of a CloudEvent as well though. I think of `instanceof` for this type of use case and an interface for shared behavior, but let's move forward as is.

My last thought from this PR is: would we want a 1.x release of the sdk to live alongside support for v1.0 of the spec and then we bump to a 2.x for v2.0 of the spec? If we go that route, we might consider removing the V1 namespace all together. I'm not too sure which would be best atm, but if 1.x of the sdk supports multiple CloudEvent spec version at the same time, I think your approach is the right one. Let's plan for supporting multiple CloudEvent spec versions atm.

---

Going to merge this so that we can keep along with implementing some of the functionality in the `V1` namespace. Thanks for this @rubenrangel!

**Comment by rubenrangel** [comment]
Sounds good @jlaswell. I think keeping CloudEvent version support decoupled from the SDK version would be useful. Since a codebase can only have one version of the library, we'd essentially be making application design decisions for developers, as their application would only be able to produce or consume one major version of the spec. In a message-driven/distributed systems world, we'd be enforcing rigidity instead of enabling flexibility.

But you're right about the `instanceof` tradeoff we'd be making, and should definitely be mindful of that.

**Pull request merged**